apps/guardian/src/pages/Actions/util.tsx

```diff
@@ -0,0 +1,267 @@
+         * Sends the contract update by accepting the proposed transaction
+         */
+        const acceptProposal = useCallback(() => {
+            if (flow === undefined) {
+                throw new Error('Expected interaction generator to still be available');
+            }
+
+            setStep(ActionStep.UpdateConctract);
+            flow.next(true)
+                .then(() => {
+                    setStep(ActionStep.Done);
+                    return sleep(1000); // Allow user to see the successful contract update step
+                })
+                .then(refreshGuardians)
+                .then(reset)
+                .catch((e: BackendError) => {
+                    setError(e.message);
+                });
+        }, [flow, reset, refreshGuardians]);
+
+        /**
+         * Reject the proposed transaction.
+         */
+        const rejectProposal = useCallback(() => {
+            if (flow === undefined) {
+                throw new Error('Expected interaction generator to still be available');
+            }
+
+            void flow.next(false);
+            reset();
+        }, [flow, reset]);
+
+        // Runs when modal shows
+        useEffect(() => {
+            if (isOpen && flow !== undefined) {
+                flow.next()
+                    .then((res) => {
+                        setProposal(res.value as P);
+                        setStep(ActionStep.HandleProposal);
+                    })
+                    .catch((e: BackendError) => {
+                        setError(e.message);
+                    });
+            }
+        }, [isOpen, flow]);
+
+        // Clean up on unmount
+        useEffect(
+            () => () => {
+                reset();
+            },
+            [reset],
+        );
+
+        const props: ActionableStepsChildProps<P> = useMemo(
+            () => ({
+                step,
+                proposal,
+                error,
+                initFlow: () => setIsOpen(true),
+                hide: reset,
+                acceptProposal,
+                rejectProposal,
+                isOpen,
+            }),
+            [acceptProposal, error, proposal, rejectProposal, isOpen, step, reset],
+        );
+
+        return <Component {...props} />;
+    };
+}
+
+type AwaitPeersProps = PropsWithChildren<{
+    predicate(g: GuardianState): boolean;
+}>;
+
+/**
+ * Shows the progress of peer registrations for a specific step in the election
+ */
+export function AwaitPeers({ predicate, children }: AwaitPeersProps) {
+    const { guardians } = useAtomValue(guardiansStateAtom);
+
+    if (guardians === undefined) {
+        return null;
+    }
+
+    const numRegistered = guardians.filter(([, g]) => predicate(g)).length;
+    const progress = numRegistered * (100 / guardians.length);
+
+    return (
+        <div>
+            <h3 className="text-center">{children}</h3>
+            <ProgressBar now={progress} label={`${numRegistered}/${guardians.length}`} />
```

Comment:
This is for my understanding, but you are guarding against guardians being undefined, but not against being an empty array. I suppose the latter will not happen if we set up the election in a non-stupid way.

Reply:
Hoping the previous comment will explain 🙏